### ISSUE TITLE
zshrc: All commands are executed in / if the directory does not exist

### ISSUE THREAD

**Issue by yschroeder** [grml/grml-etc-core#76]
Last week, I deleted my root file system (`/`) by issuing an `rm -rf *` in
a zsh with grml config. I am reporting this here, because I could not reproduce
the issue with a plain zsh (`zsh -f`) or with a plain bash.

Here is what I did:
1. Insert a USB drive/SD card into the system
2. Gnome automounts it in `/run/media/<username>/<partition>`
3. Navigate in a zsh into that folder (`cd  /run/media/<username>/<partition>`)
4. Remove the USB drive/SD card physically from the system
5. Gnome unmounts it automatically
5.1 (optional) plug the drive back in, doesn't really matter
6. Enter `pwd` into the open zsh (in the now unexisting folder)
-> output will be `/run/media/<username>/<partition>`
7. Enter`ls`
-> Prints the content of `/`
8. Enter `echo *`
-> Prints the content of `/`
9. DO NOT DO THIS ON A LIVE SYSTEM!!!: Enter `sudo rm -rf *`
-> Deletes your whole root file system

Basically zsh (with grml config) tells you in the prompt, in `pwd` and
in `$PWD` that you are in the mounted drive while all commands are
executed with `/` as working directory.

In my case the mounted drive was another linux root file system so doing
`ls` did show a root fs as expected, but it was my own one, not the one
of the USB drive.

Other shells (plain zsh and bash) raise errors on `ls` instead of
silently redirecting to `/` and also do not expand the `*` to all files in `/`.

I hope you can track down this bug, as it is extremely dangerous and
destroys all confidence in what the shell shows me...

**Comment by mika** [comment]
@yschroeder JFTR, what distribution + release and which zsh version are you using?

If I do:

```
mkdir /tmp/foo
cd /tmp/foo
rmdir /tmp/foo
```

... then `zsh -f`, `bash` as well as zsh with grml-zshrc all behave identical, reporting `/tmp/foo` as working directory (when invoking `pwd`). Also neither `echo *` nor `ls` print the content of `/` for me.
So I'm not yet sure how to reproduce this. Do you have a custom `~/.zshrc.local` or so?

@ft any ideas what might be going on and what to do about it?

**Comment by yschroeder** [comment]
I am also not able to reproduce it your way. Also I noticed, that if you `ls` right after unplugging the USB drive, it behaves correctly (shows nothing), only a few seconds later consecutive `ls` show the content of `/`. I believe that Gnome does something special when unmounting, because `umount` manually does not even work as it is still in use...

I am using Arch Linux with grml config from the package manager:
`5.0.7-arch1-1-ARCH #1 SMP PREEMPT Mon Apr 8 10:37:08 UTC 2019 x86_64 GNU/Linux`
GRML config version:
`grml-zsh-config 0.16.0-1 `
zsh version:
`zsh 5.7.1 (x86_64-pc-linux-gnu)`

I have no custom `~/.zshrc.local`. The only extra plugin I use apart from grml config is `zsh-autosuggestions`.

**Comment by yschroeder** [comment]
I found a way to reproduce it for you:

You need a source to mount and mount it into another file system. It does not work if the source is in the same file system as the destination (e.g. source and destination in `/tmp`)

```
mkdir /tmp/source
mkdir ~/destination
touch /tmp/source/test
sudo mount --bind /tmp/source ~/destination
cd ~/destination
ls # shows the file test
sudo umount --lazy ~/destination
ls # sometimes shows nothing, depending how fast you are
ls # shows / (plain zsh shows the file test)
```

It looks like zsh with grml config does not hold the handle to the directory, so `umount` unmounts it. A plain zsh seems to keep the handle and still shows the file `test` in `destination`.

**Comment by yschroeder** [comment]
Whats the status of this? Were you able to reproduce the problem?

**Comment by mika** [comment]
No one had the time yet to look into this.

Feel free to dig into this yourself, e.g. by bisecting the grml-zshrc (to identify which configuration variable/option/.... is responsible for this). Once we know what setting is responsible we might be able to even clarify this with zsh upstream.

**Comment by yschroeder** [comment]
Ok, I found out that it is in fact triggered by the `vcs_info` functionality and is not only an issue with grml config.

I used the following zshrc in a plain zsh (`zsh -f`):
```zsh
autoload -Uz vcs_info
zstyle ':vcs_info:*' enable git svn
precmd() {
    vcs_info
}

setopt prompt_subst
PROMPT='${vcs_info_msg_0_}%# '
```
The config was copied from here: http://arjanvandergaag.nl/blog/customize-zsh-prompt-with-vcs-info.html

It also triggers the bug.

However, I still do not understand why Oh My Zsh is not affected. I believe they also use the vcs functions...

**Comment by envy** [comment]
We narrowed it down even further, it has something to do with the realpath substitution in zsh itself.
Here is a new minimal example that works without any zsh config:

```
zsh -f
mkdir /tmp/source
mkdir ~/destination
touch /tmp/source/test
sudo mount --bind /tmp/source ~/destination
cd ~/destination
ls # shows the file test
sudo umount --lazy ~/destination
ls # shows the file test
export foo="."
export bar=${foo:P}
ls # shows /
```

In grml, this is triggered inside the `vcs` `precmd` during `vcs_info` when having, e.g., `svn` installed which uses the `VCS_INFO_bydir_detect` functionality.

This seems to have been introduced during this commit: https://github.com/zsh-users/zsh/commit/faa07d064bb2bd9cd9892a255a4b63811242b9fb

How do we report this upstream to zsh directly?

**Comment by mika** [comment]
Nice findings, @yschroeder and @envy - thanks!

I'd recommend mailing this to the zsh development mailing list (zsh-workers@zsh.org, see), maybe also explicitly Cc-ing the author of the commit (dana@dana.is).

**Comment by yschroeder** [comment]
We just found it also works with `:A` and `:a`.

Better minimal example instead of the two `export`s:
```
foo=. ; echo ${foo:P}
```

**Comment by Valodim** [comment]
`sudo zsh -f -c 'mount -t tmpfs tmpfs a; cd a; touch x; ls; umount --lazy ../a; ls; foo=.; echo ${foo:a}; ls; realpath .; echo $PWD;'`

**Comment by Valodim** [comment]
This happens with all of the :P, :A and :a modifiers. Tracing [:a](https://github.com/zsh-users/zsh/blob/3d7215cc8277b39cd1e24ce5a04376d45bfbabf0/Src/hist.c#L814) as the simplest of them, it does [chabspath](https://github.com/zsh-users/zsh/blob/3d7215cc8277b39cd1e24ce5a04376d45bfbabf0/Src/hist.c#L1816), which does [zgetcwd](https://github.com/zsh-users/zsh/blob/3d7215cc8277b39cd1e24ce5a04376d45bfbabf0/Src/compat.c#L490), which does [zgetdir](https://github.com/zsh-users/zsh/blob/3d7215cc8277b39cd1e24ce5a04376d45bfbabf0/Src/compat.c#L317), which does chdir stuff.

`zgetdir` is *ancient*, most of it coming from a 20 years old "initial revision" commit, and I'm having a hard time following it :) guess we should refer this to zsh-workers.

**Comment by yschroeder** [comment]
I brought this issue to the zsh-workers mailing list:
http://www.zsh.org/mla/workers/2019/msg00291.html

**Comment by mika** [comment]
Lovely, thanks for taking care! I saw that Peter Stephenson himself already replied there, I'd say that's excellent news. :) Great job, folks! :+1:

**Comment by yschroeder** [comment]
The issue got fixed here:
https://github.com/zsh-users/zsh/commit/d4972af12c29bded4706464dbae3458e5a425d00

I tested it and it got indeed fixed.
